daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/value/ValueHasher.scala

```diff
@@ -0,0 +1,173 @@
+// Copyright (c) 2019 The DAML Authors. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.digitalasset.daml.lf
+package value
+
+import java.nio.ByteBuffer
+import java.security.MessageDigest
+
+import com.digitalasset.daml.lf.data.{Numeric, Ref, Utf8}
+import com.digitalasset.daml.lf.value.Value._
+
+abstract class ValueHasher {
+
+  private def encode(a: Array[Byte]) =
+    Ref.LedgerString.assertFromString(
+      a.foldLeft(StringBuilder.newBuilder)((acc, b) => acc.append("%02x".format(b))).toString()
+    )
+
+  def hashValue(value: Value[AbsoluteContractId]): Array[Byte]
+
+  final def hashValueString(value: Value[AbsoluteContractId]): Ref.LedgerString =
+    encode(hashValue(value))
+
+  def hashContract(value: Value[AbsoluteContractId], identifier: Ref.Identifier): Array[Byte]
+
+  final def hashContractString(
+      identifier: Ref.Identifier,
+      value: Value[AbsoluteContractId]
+  ): Ref.LedgerString =
+    encode(hashContract(value, identifier))
+
+  def hashContractKey(identifier: Ref.Identifier, value: Value[AbsoluteContractId]): Array[Byte]
+
+  final def hashContractKeyString(
+      identifier: Ref.Identifier,
+      value: Value[AbsoluteContractId]
+  ): Ref.LedgerString =
+    encode(hashContractKey(identifier, value))
+
+}
+
+object ValueHasher extends ValueHasher {
+
+  // tags are used to avoid hash collisions due to equal encoding for different objects
+  private val tagUnit: Byte = 1
```

Comment:
```suggestion
  private val Seq(tagUnit, tagTrue, /*others*/, _*) =
    collection.immutable.NumericRange.inclusive[Byte](1, 127, 1)
```

Reply:
I prefer the original approach. 
1 - the tag should never change 
2 - we will have to specify the hashing method somewhere, so it is simpler if the tag are explicit.